les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// Limiter protects a network request serving mechanism from denial-of-service attacks.
+// It limits the total amount of resources used for serving requests while ensuring that
+// the most valuable connections always have a reasonable chance of being served.
+type Limiter struct {
+	lock sync.Mutex
+	cond *sync.Cond
+	quit bool
+
+	nodes                                map[enode.ID]*nodeQueue
+	addresses                            map[string]*addressGroup
+	addressSelect, valueSelect           *WeightedRandomSelect
+	maxValue                             float64
+	maxWeight, sumWeight, sumWeightLimit uint
+	selectedByValue                      bool
+}
+
+// nodeQueue represents queued requests coming from a single node ID
+type nodeQueue struct {
+	queue                   []request
+	id                      enode.ID
+	address                 string
+	value                   float64
+	flatWeight, valueWeight uint64 // current selection weights in the address/value selectors
+	nextWeight              uint   // next request weight
+	sumWeight               uint   // summed weight of requests queued by the node
+	groupIndex              int
+}
+
+// addressGroup is a group of node IDs that have sent their last requests from the same
+// network address
+type addressGroup struct {
+	nodes                      []*nodeQueue
+	nodeSelect                 *WeightedRandomSelect
+	sumFlatWeight, groupWeight uint64
+}
+
+// request represents an incoming request scheduled for processing
+type request struct {
+	process chan chan struct{}
+	weight  uint
+}
+
+// flatWeight distributes weights equally between each active network adress
+func flatWeight(item interface{}) uint64 { return item.(*nodeQueue).flatWeight }
```

Comment:
Pehaps change `weight` to `selectionProb`. I'm also not good at naming :P. Totally free for picking a better name

Reply:
see above